### PR TITLE
Fix log provider rate limit integration test

### PR DESCRIPTION
**Pull request by EasterTheBunny**
This commit introduces three main integration tests for ReadLogs rate limiting. The tests attempt to capture use cases for the rate limiting feature. As the tests are currently written, some changes were required in the provider `readLogs` function to satisfy the test cases.

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go [hunk 1]
```diff
@@ -329,46 +329,88 @@ func (p *logEventProvider) readLogs(ctx context.Context, latest int64, entries [
...
+		// added :=
+		p.buffer.enqueue(entry.upkeepID, logs...)
+
+		// fmt.Printf("added: %d; logs: %d\n", added, len(logs))
+
+		// if no new logs were added or no logs were found, update the last poll block
+		// if added > 0 || len(logs) == 0 {
+		entry.lastPollBlock = latest
+
+		p.filterStore.UpdateFilters(func(original upkeepFilter, updated upkeepFilter) upkeepFilter {
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:407]: We update the last poll block from another function

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go [hunk 2]
```diff
@@ -329,46 +329,88 @@ func (p *logEventProvider) readLogs(ctx context.Context, latest int64, entries [
...
+
+		// start should either be the last block polled for the entry or the
+		// lookback range in the case this is the first time the entry is polled
 		start := entry.lastPollBlock
 		if start == 0 || start < latest-logBlocksLookback {
 			// long range or first time polling,
 			// using a larger lookback and burst
 			start = latest - logBlocksLookback*2
+
+			// start should not be negative in the special case of an empty or
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:341]: good catch, but it should be outside of this if statement, right before we reserve with `latest-start`

Comment (EasterTheBunny) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:341]: Makes sense.

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go (hunk 1 above)
Comment (EasterTheBunny) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:407]: In that case, we also don't need `entry.lastPollBlock = latest` in the condition of `added > 0 || len(logs) == 0`?

Comment (EasterTheBunny) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go]: NVM

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go [hunk 3]
```diff
@@ -25,12 +25,12 @@ import (
 )
 
 const (
-	BlockLimitExceeded = "block limit exceeded"
-	logTriggerType     = 1
+	logTriggerType = 1
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:28]: use ocr2keepers.LogTrigger

**Review by shileiwill — DISMISSED** [review]
LGTM.

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/integration_test.go [hunk 4]
```diff
@@ -115,76 +116,227 @@ func TestIntegration_LogEventProvider(t *testing.T) {
...
+		logprovider.LogEventProviderTest,
+		[]*big.Int,
+		[]deferableFunc,
+	) {
+		ctx, cancel := context.WithCancel(testutils.Context(t))
+		backend, stopMining, accounts := setupBackend(t)
+		userContractAccount := accounts[2]
+		db := setupDB(t)
+
+		deferables := []deferableFunc{
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/integration_test.go:137]: use a single function to close everything

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/integration_test.go [hunk 5]
```diff
@@ -115,76 +116,227 @@ func TestIntegration_LogEventProvider(t *testing.T) {
...
+	t.Run("should allow constant polls within the rate and burst limit", func(t *testing.T) {
+		ctx, backend, poll, logProvider, ids, defers := setupTest(t, &logprovider.LogEventProviderOptions{
+			// BlockRateLimit is set low to ensure the test does not exceed the
+			// rate limit
+			BlockRateLimit: rate.Every(50 * time.Millisecond),
+			// BlockLimitBurst is just set to a non-zero value
+			BlockLimitBurst: 5,
+		})
+
+		defer func() {
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/integration_test.go:205]: single function as mentioned above

**Review by shileiwill — APPROVED** [review]

**Pull request merged**